File: 23/2.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    fstream file;
    file.open("2.in", ios::in);
    int ans = 0;
    if (file.is_open()) {
        string line;
        vector<unordered_map<char, int>> minCubes;
        while (getline(file, line)) {
            int pos_col = line.find(": ");
            int i=0;
            for (;line[i]!='\0';i++){}
            string cubeSets = line.substr(pos_col+2, i);
            unordered_map<char, int> um = {{'r',0}, {'g',0}, {'b',0}};
            i=0;
            string num = "";
            while (isdigit(cubeSets[i])) {num+=cubeSets[i]; i++;}
            um[cubeSets[++i]] = stoi(num);
            for (; cubeSets[i] != '\0'; ++i) {
                if (cubeSets[i] == ',' || cubeSets[i] == ';'){
                    i+=2;
                    num = "";
                    while (isdigit(cubeSets[i])) {num+=cubeSets[i]; i++;}
                    if (um[cubeSets[i+1]] < stoi(num)) um[cubeSets[i+1]] = stoi(num);
                } 
            }
            minCubes.push_back(um);
        }
        int rm=12, gm=13, bm=14;
        for (int i=0; i<minCubes.size(); ++i) {
            // Part 1            
            /* if (minCubes[i]['r']<=rm && minCubes[i]['g']<=gm && minCubes[i]['b']<=bm ) { */
            /*     ans += i+1; */
            /*     cout << "for this " << i+1 << endl; */
            /* } */
            // Part 2
            ans+=minCubes[i]['r']*minCubes[i]['g']*minCubes[i]['b'];
        }
    }
    cout << "\nans  = " << ans << "\n";
    return 0;
}

File: 23/2.py
allGames = []
with open('2.in', 'r') as f:
    for line in f.readlines():
        um = {'r':0,'g':0,'b':0}
        game = line.split(': ')[1]
        print(game)
        for c in range(len(game)):
            if game[c].isnumeric():
                num = game[c]
                c+=1
                while game[c].isnumeric():
                    num+=game[c]
                    c+=1
                um[game[c+1]] = max(um[game[c+1]], int(num))
        print("um", um)
        allGames.append(um)
print(allGames)


ans = 0
for i in range(len(allGames)):
    if allGames[i]['r'] <=12  and allGames[i]['g'] <= 13  and allGames[i]['b'] <=14:
        print("this ", i+1)
        ans+=i+1
print("___________", ans)


